### ISSUE TITLE
Potential illegible font size for map attribution and popup content

### ISSUE THREAD

**Issue by natevw** [Leaflet/Leaflet#7952]
As was foretold by https://github.com/Leaflet/Leaflet/pull/7571#issuecomment-960906568, the changes made in https://github.com/Leaflet/Leaflet/pull/7571 / https://github.com/Leaflet/Leaflet/pull/7800 can result in very small font sizes in some contexts. I.e. looks like the IE compatibility concern was addressed, but **not** the general issue of "what happens if a site has its own `html { font-size: --arbitrary-value; }`"?

For example one site I'm maintaining was started out years ago based on a Bootstrap stylesheet that makes a `rem` work out to `10px` and this led to nearly-illegible popup and attribution content after updating to the latest Leaflet. [To be clear: when the `rem` is set to `10px` or `62.5%` at the `html` level, usually the initial `em` font size for the page is then scaled back **up** at the `body` level. I'm not complaining about the page **and** Leaflet having too-small text — only Leaflet ends up with tiny text in practice! See example in my later comment below.]

~~In my case I simply worked around by adding a subset of this~~

```
// [old] WORKAROUND: 
.leaflet-container { font-size: 12px; }
.leaflet-container .leaflet-control-attribution,
.leaflet-container .leaflet-control-scale,
.leaflet-control-scale-line { font-size: 11px; }
```

**UPDATE**: now after https://github.com/Leaflet/Leaflet/pull/8057 the workaround should need only:

```
// set up units to match Leaflet expectation
.leaflet-container { font-size: 1.2rem; }
```

to the site's own stylesheet which is included after the vendored Leaflet one and so takes precedence.

**Comment by Falke-Design** [comment]
Can you please create a sample www.leafletjs.com/edit.html

**Comment by Malvoz** [comment]
What do you mean by "illegal font size"? [Most countries apply laws and policies based on WCAG](https://www.w3.org/WAI/policies/). But WCAG does not have a "minimum font size required" success criterion. Though there is https://www.w3.org/TR/WCAG21/#resize-text which can't reliably be met using `px` units, as I understand it. See references:

- https://www.craigabbott.co.uk/blog/accessibility-and-font-sizes
- https://www.24a11y.com/2019/pixels-vs-relative-units-in-css-why-its-still-a-big-deal/<br>
  <img width="500" src="https://user-images.githubusercontent.com/26493779/151317439-9901e985-b670-4fe2-84f8-d9462812db25.png">

I guess https://github.com/Leaflet/Leaflet/pull/7800 was merged too early, considering it was concluded to be a breaking change in https://github.com/Leaflet/Leaflet/pull/7571#issuecomment-960906568?

**Comment by natevw** [comment]
@Falke-Design The easiest way to see it is pasting a block like the following to e.g. "debug/map/map-popup.html" or "debug/map/popup.html" after the usual stylesheet reference:

```
<style>
  html { font-size: 10px; }
  body { font-size: 16px; }
</style>
```

<img width="1000" alt="Example of normal looking map but popup contents and attribution text are tiny" src="https://user-images.githubusercontent.com/265902/151448619-7fb5c745-340f-4c1b-bf9b-64af30794abb.png">

Of course real-world examples wouldn't be quite so simplistic, but even in this form it's ± a fair representation of what Bootstrap (3.x) at least used to do, presumedly so that `rem` could be used as a convenient decimal number of pixels.

**Comment by Malvoz** [comment]
> looks like the IE compatibility concern was addressed, but not the general issue of "what happens if a site has its own `html { font-size: --arbitrary-value; }`"?

That's not an issue with Leaflet, that's an issue with small font sizes on web pages.

If I have a generic HTML document and add `<small>© Brand 2022</small>`, that will be smaller than say `<p>` or `<h1>`. If I then set the document's font-size to 10px, `<small>` will be even smaller.

Similar to the `<small>` element - attribution etc. are intended to be subtle - but there's a responsibility of authors to use legible font-sizes. I'd argue that a 10px font-size on the document is bad practice.

**Comment by natevw** [comment]
> That's not an issue with Leaflet, that's an issue with small font sizes on web pages.

I might update my original post since a key context here got a bit buried downthread. The rest of the page content is **not** small. Again, the trick some developers like to use is to configure `rem` as a particular baseline without actually changing the initial `em` size e.g.:

```
<style>
  html { font-size: 10px; }
  body { font-size: 16px; }
</style>
```

Is that a **good** thing to do? I don't know. But it *is* "a thing" that gets done — and a few Leaflet components are now dependent on whatever the `rem` size happens to be on any particular page.

It seems Leaflet assumes that `1rem` will be `16px` **or larger**. But it's not unusual for developers to make `1rem` default to `10px` for convenience ([62.5%](https://www.aleksandrhovhannisyan.com/blog/respecting-font-size-preferences-rems-62-5-percent/#setting-the-base-font-size-to-625percent) perhaps even more common) and that makes font sizes of `0.75rem` and `0.69rem` quite small.

I wish I had an easy alternative, but I'm not sure what the best solution is here either. I appreciate the efforts that are being made to improve the default accessibility of Leaflet, and perhaps the changes made are the best solution overall. Or maybe someone else knows of better techniques?

In the end I'm not opposed to this issue being closed as "By Design / Won't Fix" but wanted to a least re-raise it as a breaking change and document a potential workaround for other users it might effect.

**Comment by natevw** [comment]
> I wish I had an easy alternative

Actually, maybe there is one: what if instead of assuming anything about `1rem` at the root of the document, Leaflet were to assume a reasonable `1em` just on the map container?

That might also break some websites, but imo it's a better long-term situation. That provides compatibility with:

```
html { /* font-size left at UA default */ }
```

and:

```
html { font-size: 62.5%; }
body { font-size: 1.6rem; }
```

and even in the craziest contexts the developer can easily re-set their Leaflet containers' baseline without having to rework units across their entire page:

```
html { font-size: 1px; }
body { font-size: 5000rem; }
p { font-size: 0.005em; }

.leaflet-map { font-size: 16px; }
```

This would also have the benefit of making it easy to customize the map in general, again **regardless** of relative/root units anywhere else:

```
.leaflet-map { font-size: 16px; }
.leaflet-map.compact { font-size: 14px; }
.leaflet-map.jumbo { font-size: 20px; }
```

**Comment by Malvoz** [comment]
To simplify it, if I may, there are 2 choices for Leaflet:

1. Set fixed font-sizes (previous approach, `px`)
   
   - For end-users: The primary issue with this approach is that the end-user loses control over the font size. `px` does not respect the user's font size settings in the browser, because it is a fixed font size.
   -  For developers: Depending on the font size used in the application this may be too large or too small, in that case CSS changes are required.

2. Set relative font-sizes (current approach, `rem`)
  
    - For end-users: Users can change the font size of text in documents (including text in Leaflet maps) because it is relative.
    - For developers: Same as above

@natevw is that a fair assessment?

**Comment by natevw** [comment]
@Malvoz I'm seeing at least **three** options:

1. Fixed font sizes (`px` as Leaflet v1.7)
2. Font sizes relative to **root** element (`rem` as current trunk)
3. Font sizes relative to **map** element (`em` as https://github.com/Leaflet/Leaflet/issues/7952#issuecomment-1030333605 proposes)

For option 3 I think the benefits might be:

* For end-users: still follows overall font size preference [assuming the rest of the page does]
* For developers: fairly neutral to other unit conventions they use in rest of page, and allows adjusting map elements bigger/smaller as a whole when desired

The one catch is that for both Leaflet v1.7 and current trunk it doesn't matter what the `em` happens to be at the `.leaflet-container` right now. In some pages the map might be in a part of the document where an `em` is more or less than `16px`.  **But** it's easy to reset `em` to a good value anywhere, whereas `rem` is something that **every** element has to share whether a plugin like this or website's own content.

**Comment by Malvoz** [comment]
@natevw Ah yes, I meant 2 options as in relative vs fixed. Other relative units can be considered too of course. Are there any caveats with `em` that you know of?

**Comment by Malvoz** [comment]
An alternative solution to using `em` as proposed here is suggested in https://github.com/Leaflet/Leaflet/pull/8052. To my understanding they work precisely the same, the difference is that https://github.com/Leaflet/Leaflet/pull/8052 uses less code, and does not decrease the default font size in maps in general (https://github.com/Leaflet/Leaflet/issues/8051).

**Comment by natevw** [comment]
As I mentioned in https://github.com/Leaflet/Leaflet/pull/8057#issuecomment-1067419947 and updated my OP here a bit as well, I think we are close to a good solution:

The Leaflet CSS now only uses `rem` in **one** place — the `.leaflet-container` — and everything else is relative to *that* (or perhaps still static, didn't really audit).

So this means:

1. Original goal of relative units instead of any fixed `px` is attained ❇️
2. Does still lead to regression in websites where `rem` is set  "cleverly" like `10px` tricks 🔶
3. **But** it is easier to fix, authors only have to set `.leaflet-container { font-size: 1.2rem; }` or whatever to get to accessibly adjustable 12px starting *equivalent*

So I'm leaving this open because fact remains that next release may ~~break~~ change size on some sites. If you wish to avoid that, a stopgap would be to still hardcode:

```
.leaflet-container {
  font-size: 12px;
}
```

i.e. leave all the other work in place — the relative units on things *inside* the container — but not actually gain accessibility benefits by default until a **major** release. Could encourage authors to set `.leaflet-container { font-size: 0.75rem; }` themselves until then.

**Comment by natevw** [comment]
Sorry, I ramble :-) The tl;dr; of above is now after #8057 maybe options for a *minor* release are:

* close this as WONTFIX known edge case, document workaround for sites that break
* keep all font/accessibility improvements *except* still hardcode `12px` on container
